### DISCUSSION TITLE
Disk named `public` cannot be accessed

### DISCUSSION THREAD

**Discussion by Kokil** [spatie/laravel-medialibrary#3305] [Q&A]
I'm encountering an issue while using laravel-medialibrary version ^10.9 . 
When i run php artisan media-library:regenerate it give: 

Media id 1: `Disk named `public` cannot be accessed`

I have verified my filesystem configuration in config/filesystems.php and it appears to be set correctly:

```
'public' => [
    'driver' => 'local',
    'root' => storage_path('app/public'),
    'url' => env('APP_URL').'/storage',
    'visibility' => 'public',
    'throw' => false,
],
```
Could you please guide me in troubleshooting this issue? Is there anything specific I should be aware of when using laravel-medialibrary with Laravel 10.x?

**Comment by parallels999** [comment]
Folder permissions

**Comment by Kokil** [comment]
> Folder permissions

I don't think it's due to folder permissions. 
after your comment, I tried giving even 777 to the storage folder.

If I upload from the backend it's working.

**Comment by francoism90** [comment]
@Kokil What is your environment? Is this Laravel Sail or production?

**Comment by Kokil** [comment]
> @Kokil What is your environment? Is this Laravel Sail or production?

It's in the local environment. i haven't tested that in production.

**Comment by francoism90** [comment]
@Kokil I also think this is a permission issue.

**Comment by Kokil** [comment]
![image](https://github.com/spatie/laravel-medialibrary/assets/481786/b7ad8b67-b2ce-4974-83a1-19aaed72e867)
Any Suggestion?

**Comment by francoism90** [comment]
I really don't know, if your public correctly defined in `config/filesystems.php`? Tried clearing your cache? Also does the symlink correctly exists?

Anyway, I don't think this is a package issue, sorry.

**Comment by Kokil** [comment]
After your comment, I take the following step :
1. Get the fresh filesystems.php.
2. I run artisan migrate:fresh --seed
3. cache: clear, remove old symlink, regenerate symlink
4. i upload the media from backend upload success.
5. after that i run : php artisan media-library:regenerate
again same
![image](https://github.com/spatie/laravel-medialibrary/assets/481786/f7e9b61c-d614-4304-93a9-412418719560)

PHP 8.1.7 & Laravel Framework 10.11.0
Note :
I migrate this application laravel 8 to laravel 10.
at laravel 8 i was working fine.

**Comment by bhojkamal** [comment]
I also got same error on laravel 10. but it was working in same way in laravel 9. The error is 
```
"message": "Disk named `public` cannot be accessed",
    "exception": "Spatie\\MediaLibrary\\MediaCollections\\Exceptions\\DiskCannotBeAccessed",
```

**Comment by Kokil** [comment]
Yes i also upgrade from laravel 8 in laravel 8 it was working fine.

**Comment by Kokil** [comment]
any update on this ?

**Comment by fredsal** [comment]
@Kokil Hi, make a debug, and feel free to make a PR with the fix

[laravel/compare/v8.x...v10.x](https://github.com/laravel/laravel/compare/v8.0.0...v10.2.3)

**Comment by middaparka** [comment]
I'm also seeing this issue, but only on PHP 8.2(.8) - 8.1(.12) is behaving as expected. (Both running Laravel 10.15.0.)

Specifically the `file_put_contents` call in the `writeToFile` method within `…/laravel/vendor/league/flysystem-local/LocalFilesystemAdapter.php` (line ~123) is returning false and triggering the exception, although the `error_get_last()` call response is null.

I'm assuming there must be a PHP CLI configuration difference that's causing this - did anyone else determine what their issue was?

**Reply by middaparka** [reply]
As an update, I can't see any significant PHP CLI config differences that could trigger this.

(Unless some _really_ weird things are happening in the absence of an ODBC extension.) 🤔

To clarify, this is simply switching between the active PHP CLI version on the same VM instance, so the CLI commands are running as the same user in the same environment.

**Reply by titantwentyone** [reply]
Just some more info but have come across this using Sail and PHP 8.1.20 and Laravel 10.19. I am automating the import of some media. Oddly, The first 3 images upload fine, but the fourth stops with this same error. I did this multiple times with the same result. I even changed out the order of the imports to see if it was an image that was causing the issue but no joy.

**Comment by nocodelab** [comment]
Hello guys, did anyone solve this issue? I'm having the same problem with L10 and latest version of MediaLibrary.

**Reply by middaparka** [reply]
Not that I'm aware of. 😔

I still suspect it's due to some PHP change between PHP 8.1 and 8.2, perhaps relating to how streams are handled when executing via CLI rather than via FPM, but that's just a hunch.

The fact that it only occurs during CLI/Artisan command execution also feels significant, but it's not an obvious permissions issue. Having compared my PHP FPM and CLI configs in detail I couldn't see any meaningful differences that would cause this issue.

**Reply by nocodelab** [reply]
I confirm that it happens just via CLI :/

**Reply by Kokil** [reply]
Yes i also faced in CLI only.

**Reply by nocodelab** [reply]
I don't know if this can help, but I tried to change the "MEDIA_DISK" to public and everything works well.
If I change the disk to another one (no matter if s3 or with local driver) the issue magically appears.
Did you also try that?

For instance, here is the secondary local filesystem which triggers the error:
```

        'images' => [
            'driver' => 'local',
            'root' => storage_path('app/public/medias'),
            'url' => env('APP_URL').'/medias',
            'visibility' => 'public',
            'throw' => false,
        ],
```

**Comment by PerikiyoXD** [comment]
Try setting 'throw' to true and send the exception it throws. This is just to debug, change it back to 'false' after testing.

**Reply by PerikiyoXD** [reply]
In my case it was a weird naming problem using `<filename>.png&v=<stamp>&_=<stamp>`, which is invalid extension with all the extra url stuff.

**Reply by francoism90** [reply]
Could this be an issue with the filename sanitizer?

We also had this issue, but could indeed see the error with `throw => true`.

**Reply by middaparka** [reply]
Sadly the error I'm seeing is a generic unable to write to file message, which is strange as a zero byte file is being successfully created.

I still wonder if this is somehow related to the way streams work when using CLI as opposed to "normal" web access.

**Reply by PerikiyoXD** [reply]
Maybe CLI and FPM users differ. I've happened to face this issue with write-permissions on folders.

**Comment by liamseys** [comment]
Has anyone fixed this issue already?

**Reply by PerikiyoXD** [reply]
Did you try setting `'throw' => true,` and debug from there?

**Reply by liamseys** [reply]
```
 's3' => [
            'driver' => 's3',
            'key' => env('AWS_ACCESS_KEY_ID'),
            'secret' => env('AWS_SECRET_ACCESS_KEY'),
            'region' => env('AWS_DEFAULT_REGION'),
            'bucket' => env('AWS_BUCKET'),
            'url' => env('AWS_URL'),
            'endpoint' => env('AWS_ENDPOINT'),
            'use_path_style_endpoint' => env('AWS_USE_PATH_STYLE_ENDPOINT', false),
            'throw' => false,
        ],

        'media' => [
            'driver' => 's3',
            'key' => env('AWS_ACCESS_KEY_ID'),
            'secret' => env('AWS_SECRET_ACCESS_KEY'),
            'region' => env('AWS_DEFAULT_REGION'),
            'bucket' => env('AWS_BUCKET'),
            'url' => env('AWS_URL'),
            'endpoint' => env('AWS_ENDPOINT'),
            'use_path_style_endpoint' => env('AWS_USE_PATH_STYLE_ENDPOINT', false),
            'throw' => false,
            'visibility' => 'public',
            'root' => 'media',
        ],
```

Only thing I did was another media disk set it to public and changed the root to 'media'. I'm trying to upload it to this disk from Spatie's Media Library.

**Reply by PerikiyoXD** [reply]
Provide more context. 'public' disk is the issue in this thread. I don't know if you're using S3 or other services and I recommend enabling "throw" => "true" to try diagnosing the issue more in depth.

**Comment by alhelwany** [comment]
It's a folder permission issue.

I ran this and it worked

`sudo chgrp -R www-data storage; sudo chmod -R ug+rwx storage`

Laravel 10

**Comment by wanzkarat** [comment]
I had the same error when adding new images (not regenerate) and on my side it was due to a wrong filename I tried to set (set the file extension incorrectly). Nothing to do with permissions on my end

**Comment by rohjay** [comment]
I was having this issue running laravel locally and trying to seed the db + drop in some test logo images. After doing the `chgrp` as suggested by @GeorgeTheNerd and that not having fixed the issue, I ran my command as `www-data` like this:

```
> sudo -u www-data php artisan migrate:fresh --seed
```

That did the trick :+1: hope this helps a fellow wayward dev =]

**Comment by jarbasmedeiros** [comment]
@Kokil  Go to the filesystems file and change the public disk, comment 'visibility' => 'public' and add 'ACL' => 'public-read' to the array. it will stay like this:

```
'public' => [
            'driver' => 'local',
            'root' => storage_path('app/public'),
            'url' => env('APP_URL').'/storage',
            'ACL' => 'public-read',
            //'visibility' => 'public',
            'throw' => false,
        ],
```
It works for me : D
